docs/developer/explanations/service-management-internals.md

```diff
@@ -0,0 +1,267 @@
+the application's service secret is modified.
+
+Everything is performed on the server side, the client issues only an
+`AppCreate` call containing all the necessary information.
+
+TODO: replace by equivalent SVG graphic
+```
+user --> client  :service bind S A
+         client --> server :POST AppCreate O (A, S...)
+                    server --> cluster :create app resource
+                    server <-- cluster :ok/fail
+                    //
+                    server --> cluster :read A's service secret
+                    server <-- cluster :ok/fail
+                    server --> cluster :modify A's service secret (add S...)
+                    server <-- cluster :ok/fail
+                    server --> cluster :write A's service secret
+                    server <-- cluster :ok/fail
+                    //
+         client <-- server  :report ok/fail
+user <-- client  :report ok/fail
+```
+
+### Semantics: `app push --bind S,... ... A`
+
+The named services S... are bound to the named application `A`.
+
+For a newly created application the services are bound via the
+`AppCreate` API call, and the bindings are then picked up by the
+deployment stage, for integration into the application's deployment
+resource.
+
+In the case of an already existing application the creation will fail
+and trigger a call to `AppUpdate` instead which updates the
+application's service resource with the new services.
+
+__Attention__ Note that the above does __not__ re-start the
+application. The new services apply only to the new revision of the
+application, and not the currently running revision.
+
+Integration of the new services happens then as for a new application,
+as part of its deployment, after re-staging.
+
+TODO: replace by equivalent SVG graphic
+```
+user --> client :app push ... -b S ... A
+         client --> server :POST AppCreate O (A, S...)
+                    server --> cluster :create app resource
+                    server <-- cluster :ok/fail
+         client <-- server :ok/fail
+
+	 // if create failed
+         client --> server :POST AppUpdate O A (S...)
+                    server --> cluster :modify app resource     /(instances?!)
+                    server <-- cluster :ok/fail
+                    //
+                    server --> cluster :read A's service secret
+                    server <-- cluster :ok/fail
+                    server --> cluster :modify A's service secret (set S...)
+                    server <-- cluster :ok/fail
+                    server --> cluster :write A's service secret
+                    server <-- cluster :ok/fail
+         client <-- server :ok/fail
+	 // endif
+
+         ((upload sources))
+         ((stage sources))
+         ((deploy image (stage result)))
+                server side -- per S: ((bind sequencing))
+
+user <-- client :report ok fail
+```
+
+### Semantics: `app delete A`
+
+Deletes the named application `A`.
```

Comment:
and the services secret I guess (if one exists)?

Reply:
Yes. That is one of the earliest changes in the (now collapsed) history - Make the service binding resources owned by the app, and auto-deleted. This is turn deletes the secrets.

Reply:
See next sentence, which implies this.